File: src/main/java/br/unifor/costify/application/usecase/RegisterRecipeUseCase.java
package br.unifor.costify.application.usecase;

import br.unifor.costify.application.contracts.RecipeRepository;
import br.unifor.costify.application.dto.command.RegisterRecipeCommand;
import br.unifor.costify.application.dto.entity.RecipeDto;
import br.unifor.costify.application.errors.RecipeAlreadyExistsException;
import br.unifor.costify.application.factory.RecipeFactory;
import br.unifor.costify.application.service.IngredientLoaderService;
import br.unifor.costify.application.validation.ValidationService;
import br.unifor.costify.domain.entity.Ingredient;
import br.unifor.costify.domain.entity.Recipe;
import br.unifor.costify.domain.service.RecipeCostCalculationService;
import br.unifor.costify.domain.valueobject.Id;
import br.unifor.costify.domain.valueobject.Money;
import br.unifor.costify.domain.valueobject.RecipeCost;
import br.unifor.costify.domain.valueobject.RecipeIngredient;

import java.util.Map;

public class RegisterRecipeUseCase {
  private final RecipeRepository recipeRepository;
  private final IngredientLoaderService ingredientLoaderService;
  private final RecipeFactory recipeFactory;
  private final ValidationService validationService;
  private final RecipeCostCalculationService costCalculationService;

  public RegisterRecipeUseCase(
      RecipeRepository recipeRepository,
      IngredientLoaderService ingredientLoaderService,
      RecipeFactory recipeFactory,
      ValidationService validationService,
      RecipeCostCalculationService costCalculationService) {
    this.recipeRepository = recipeRepository;
    this.ingredientLoaderService = ingredientLoaderService;
    this.recipeFactory = recipeFactory;
    this.validationService = validationService;
    this.costCalculationService = costCalculationService;
  }

  public RecipeDto execute(RegisterRecipeCommand command) {
    validationService.validateRecipeData(command.name(), command.ingredients());
    
    validateRecipeDoesNotExist(command.name());

    // Load ingredients and calculate cost
    Map<Id, Ingredient> ingredientMap = ingredientLoaderService.loadIngredients(command.ingredients());
    
    // Create temporary recipe with zero cost for cost calculation
    Recipe tempRecipe = recipeFactory.create(command.name(), command.ingredients(), Money.zero());
    RecipeCost recipeCost = costCalculationService.calculateCost(tempRecipe, ingredientMap);
    
    // Create final recipe with calculated total cost
    Recipe recipe = recipeFactory.create(command.name(), command.ingredients(), recipeCost.getTotalCost());

    Recipe savedRecipe = recipeRepository.save(recipe);

    return RecipeDto.from(savedRecipe);
  }

  private void validateRecipeDoesNotExist(String name) {
    if (recipeRepository.existsByName(name)) {
      throw new RecipeAlreadyExistsException("Recipe with name '" + name + "' already exists");
    }
  }

}

File: internal/application/usecase/register_recipe_usecase.go
package usecase

import (
	"github.com/vini/costify-go/internal/application/contract"
	"github.com/vini/costify-go/internal/application/dto/command"
	"github.com/vini/costify-go/internal/application/dto/entity"
	"github.com/vini/costify-go/internal/application/errors"
	"github.com/vini/costify-go/internal/application/factory"
	"github.com/vini/costify-go/internal/application/service"
	domainservice "github.com/vini/costify-go/internal/domain/service"
	"github.com/vini/costify-go/internal/domain/valueobject"
)

// RegisterRecipeUseCase handles the registration of new recipes
type RegisterRecipeUseCase struct {
	recipeRepository         contract.RecipeRepository
	ingredientLoaderService  *service.IngredientLoaderService
	recipeFactory            *factory.RecipeFactory
	costCalculationService   *domainservice.RecipeCostCalculationService
}

// NewRegisterRecipeUseCase creates a new RegisterRecipeUseCase
func NewRegisterRecipeUseCase(
	recipeRepository contract.RecipeRepository,
	ingredientLoaderService *service.IngredientLoaderService,
	recipeFactory *factory.RecipeFactory,
	costCalculationService *domainservice.RecipeCostCalculationService,
) *RegisterRecipeUseCase {
	return &RegisterRecipeUseCase{
		recipeRepository:        recipeRepository,
		ingredientLoaderService: ingredientLoaderService,
		recipeFactory:           recipeFactory,
		costCalculationService:  costCalculationService,
	}
}

// Execute executes the register recipe use case
func (uc *RegisterRecipeUseCase) Execute(cmd command.RegisterRecipeCommand) (entity.RecipeDto, error) {
	// Validate that recipe doesn't already exist
	exists, err := uc.recipeRepository.ExistsByName(cmd.Name)
	if err != nil {
		return entity.RecipeDto{}, err
	}
	if exists {
		return entity.RecipeDto{}, errors.NewRecipeAlreadyExistsError(
			"Recipe with name '" + cmd.Name + "' already exists")
	}

	// Convert command ingredients to value objects
	var recipeIngredients []valueobject.RecipeIngredient
	for _, cmdIng := range cmd.Ingredients {
		id := valueobject.Id{}.Of(cmdIng.IngredientId)
		unit, exists := valueobject.FromString(cmdIng.Unit)
		if !exists {
			return entity.RecipeDto{}, errors.NewInvalidUnitError("Invalid unit: " + cmdIng.Unit)
		}
		
		recipeIngredient, err := valueobject.NewRecipeIngredient(id, cmdIng.Quantity, unit)
		if err != nil {
			return entity.RecipeDto{}, err
		}
		recipeIngredients = append(recipeIngredients, recipeIngredient)
	}

	// Load ingredients and calculate cost
	ingredientMap, err := uc.ingredientLoaderService.LoadIngredients(recipeIngredients)
	if err != nil {
		return entity.RecipeDto{}, err
	}

	// Create temporary recipe with zero cost for cost calculation
	zeroMoney := valueobject.Money{}.Zero()
	tempRecipe, err := uc.recipeFactory.Create(cmd.Name, recipeIngredients, zeroMoney)
	if err != nil {
		return entity.RecipeDto{}, err
	}

	// Calculate recipe cost
	recipeCost, err := uc.costCalculationService.CalculateCost(tempRecipe, ingredientMap)
	if err != nil {
		return entity.RecipeDto{}, err
	}

	// Create final recipe with calculated total cost
	recipe, err := uc.recipeFactory.Create(cmd.Name, recipeIngredients, recipeCost.TotalCost())
	if err != nil {
		return entity.RecipeDto{}, err
	}

	// Save recipe
	savedRecipe, err := uc.recipeRepository.Save(recipe)
	if err != nil {
		return entity.RecipeDto{}, err
	}

	return entity.FromRecipe(savedRecipe), nil
}